ios/CoinbaseWalletSDK/CoinbaseWalletSDK.swift

```diff
@@ -153,38 +147,43 @@ public final class CoinbaseWalletSDK {
                 return
             }
             
-            self.taskManager.registerResponseHandler(for: request, onResponse)
+            TaskManager.registerResponseHandler(for: request, host: self.host, onResponse)
         }
     }
     
     // MARK: - Receive message
     
-    private func isWalletSegueMessage(_ url: URL) -> Bool {
-        return url.host == callback.host && url.path == callback.path
+    static private func isWalletSegueMessage(_ url: URL) -> Bool {
+        return url.host == CoinbaseWalletSDK.callback!.host && url.path == CoinbaseWalletSDK.callback!.path
     }
 
     /// Handle incoming deep links
     /// - Parameter url: deep link url
     /// - Returns: `false` if the input was not response message type, `true` if SDK handled the input, or throws error if it failed to decode response.
     @discardableResult
-    public func handleResponse(_ url: URL) throws -> Bool {
+    static public func handleResponse(_ url: URL) throws -> Bool {
         guard isWalletSegueMessage(url) else {
             return false
         }
         
-        let response = try decodeResponse(url)
-        taskManager.runResponseHandler(with: response)
+        let encryptedResponse: EncryptedResponseMessage = try MessageConverter.decodeWithoutDecryption(url)
+        let request = TaskManager.getHost(for: encryptedResponse)
+        guard let instance = instances[request!] else {
+            throw Error.walletInstanceNotFound
+        }
+
+        let response = try instance.decodeResponse(url, encryptedResponse)
+        TaskManager.runResponseHandler(with: response)
         return true
     }
     
-    private func decodeResponse(_ url: URL) throws -> ResponseMessage {
+    private func decodeResponse(_ url: URL,_ encryptedResponse: EncryptedResponseMessage) throws -> ResponseMessage {
```

Comment:
let's use either one of them
```suggestion
    private func decodeResponse(_ EncryptedResponseMessage) throws -> ResponseMessage {
```

Reply:
for this, not sure what we can do, as I need the url to get the symmetricKey

Reply:
then let's refactor `MessageConverter.decode` to take `EncryptedContent` instead. we can address it on follow up